fdbserver/TLogServer.actor.cpp

```diff
@@ -2256,9 +2256,9 @@ ACTOR Future<Void> tLogCommit(TLogData* self,
 	if (req.debugID.present())
 		g_traceBatch.addEvent("CommitDebug", tlogDebugID.get().first(), "TLog.tLogCommit.After");
 	if (SERVER_KNOBS->ENABLE_VERSION_VECTOR_TLOG_UNICAST) {
-		self->unknownCommittedVersions.push_front(req.version);
+		self->unknownCommittedVersions.push_front(std::make_tuple(req.version, req.tLogCount));
 		while (!self->unknownCommittedVersions.empty() &&
-		       self->unknownCommittedVersions.back() <= req.knownCommittedVersion) {
+		       std::get<0>(self->unknownCommittedVersions.back()) <= req.knownCommittedVersion) {
```

Comment:
Is this correct? I assume that we are not changing the way we compute "KnownCommittedVersion" on commit proxies.

Here is what the paper says about how KCV is updated on commit proxies:

> LogServers reply to the Proxy once the log data is made durable, and the Proxy updates its KCV to the LSN if all replica LogServers have replied and this LSN is larger than the current KCV.

I think this logic doesn't work with UNICAST. Consider this scenario:

Shards A, B, and C. Let's assume tlog "tl_a" serves shard A, tlog "tl_b" serves shard B, and tlog "tl_c" serves shard C.

- Proxy, pr_a, is committing version 400 that mutated shards A and B. TLog, tl_a, makes that version durable and replies to the proxy. Tlog "tl_b" hasn't sent a reply to the proxy yet.
- Proxy, pr_b, is committing version 500 that mutated shard C. TLog, tl_c, makes that version durable and replies to the proxy, and proxy pr_b updates its KCV to 500.
- Proxy, pr_b, tries to commit version 800 that mutated shard A.
- Tlog "tl_a" receives the request and prunes version 400 from "unknownCommittedVersions".

Let's assume there is a recovery at this point.

- Tlog "tl_a" sends "<800, 1>" (with KCV = 500 and DV = 500) to the sequencer.
- TLog "tl_b" sends "<400, 2>" (with a KCV that is below 400 and DV = 400) to the sequencer.
- Tlog "tl_c" sends "<500, 1>" (with a KCV that is below 400 and DV = 500) to the sequencer.

The sequencer follows this algorithm:

> The version list is enumerated and, for each version, a count of the number of times the version appears across all tLog's unknownCommittedVersions is found. For each tLog, the highest version that is >= # tlogs in transaction - replicat count + 1 is determined and stored in the map rvLogs[tlog id].
> 
> The variable "maxRv" is then set to the maximum value in the rvLogs set.
> 
> In newEpoch(), the interface "initializeTLog" starts each tLog. It will include the rvLogs[tlog id] and maxRv.
> 

So the sequencer sets rvLogs[tl_a] to 800, rvLogs[tl_b] to a value that is below 400, rvLogs[tl_c] to 500, and mzxRv to 800, which is not correct since tlog tl_a is including version 400 and tlog tl_b is excluding version 400 (so we have a broken transaction).

Reply:
I also thought about this. From what I read from commit proxy code, it updates `KnownCommittedVersion` after reporting the commit version to the master: https://github.com/jzhou77/foundationdb/blob/14727555dc4744559669834384a872e67c086e39/fdbserver/CommitProxyServer.actor.cpp#L1431-L1449. We may need to add `SERVER_KNOBS->ENABLE_VERSION_VECTOR` condition to line 1431.

So this `KnownCommittedVersion` implies "waitForPrev" has been used at the mater, thus is consistent with previous behavior, i.e., all versions before have been committed.

Reply:
I do not think the KCV would be set to 500 in pr_b in this scenario, because the KCV is updated after ReportRawCommittedVersionRequest. And the sequencer waits for 400 before replying to the client, and only after replying to the client shall it reply to the proxy.

Reply:
> We may need to add SERVER_KNOBS->ENABLE_VERSION_VECTOR condition to line 1431.

Yes, that will need to be added. Agreed.

Reply:
Wait, don't we still have the same problem even if KCV is updated after ReportRawCommittedVersionRequest? In the above example, proxy pr_b updates its KCV to 500 only after proxy pr_a reports (to the sequencer) that version 400 has been committed, that's fine, but version 400 could still be in the unknownCommittedVersions list on tlog tl_b and tlog tl_a could have updated its KCV to 500 at the time of recovery. So the sequencer could still send these lists to the sequencer during recovery:

- Tlog "tl_a" sends "<800, 1>" (with KCV = 500 and DV = 800) to the sequencer.
- TLog "tl_b" sends "<400, 2>" (with a KCV that is below 400 and DV = 400) to the sequencer.
- Tlog "tl_c" sends "<500, 1>" (with a KCV that is below 400 and DV = 500) to the sequencer.

And, the sequencer will set rvLogs[tl_a] to 800, rvLogs[tl_b] to a value that is below 400?
Do you agree?

Reply:
it seems possible. If we know all entries in the list less than max(KCV) have been committed (per `waitForPrev`), then a solution may be to recover any entries less than the KCV that have not been pruned.

Reply:
> a solution may be to skip entries in the unknownCommittedVersions list less than the KCV that have not been pruned.

You mean less than "max(KCV)"?

Reply:
Yes (Discussed offline)

Reply:
Please update the recovery algorithm document accordingly. Thanks!